### ISSUE TITLE
Planetscale adapter is not picking up the correct database table

### ISSUE THREAD

**Issue by Janushan** [prisma/prisma#21511]
### Bug description

For my application I use a multi-tenant database system with planetscale and prisma.
This means I do not have 1 database connection url to access all of my databases at once. Rather, I need to use a separate connection string for each database.

To accommodate this use-case I have been setting up the connection strings like this up until now:
```
const prisma = new PrismaClient({
    datasources: {
      db: {
        url: databaseUrl,
      },
    },
})
```

This allows me to create a new PrismaClient for each database tenant and specify a databaseUrl for each one.

However, when I tried to replicate this with the Planetscale adapter like so:
```
  const connection = connect({ 
    url: databaseUrl 
  });
  const adapter = new PrismaPlanetScale(connection);

  const prisma = new TenantPrismaClient({
    adapter
  })
```

I get this error:
```
DatabaseError: target: dev-DB_NAME-db.-.primary: vttablet: rpc error: code = NotFound desc = Table 'mysql.TABLE_NAME' doesn't exist (errno 1146) (sqlstate 42S02) (CallerID: 5nygsvsauz06gm2xp5d2): Sql: "select count(*) from (select TABLE_NAME.id from mysql.TABLE_NAME where :vtg1 /* INT64 */ = :vtg1 /* INT64 */) as sub", BindVars: {REDACTED}
    at Connection.execute (PATH/node_modules/@planetscale/database/dist/cjs/index.js:99:19)
    at processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async PrismaPlanetScale.performIO (PATH/node_modules/@prisma/adapter-planetscale/dist/index.js:143:22)
    at async PrismaPlanetScale.queryRaw (PATH/node_modules/@prisma/adapter-planetscale/dist/index.js:114:54)
    at async PATH/generated/prisma-tenant/client/runtime/library.js:26:4983
```

I am using
```
    "prisma": "^5.4.2",
    "@prisma/client": "^5.4.2",
```

I have contacted Planetscale support as well and we suspect it might be a problem with the adapter because:
- the connection is working correctly
- the database name is not being picked up correctly
- the database name is defaulting to the default `mysql` one

I have tested the database connection string separately and have confirmed it works without the Planetscale adapter.

### How to reproduce

1. Manually specify a different database connection string from the one in the .env
2. Create a new PrismaClient from this
3. Try to connect to the different database and do a simple query

### Expected behavior

Output the expected response from the query without errors.

### Prisma information

```
generator client {
  provider        = "prisma-client-js"
  binaryTargets   = ["native", "darwin", "darwin-arm64", "linux-musl-openssl-3.0.x", "linux-musl-arm64-openssl-3.0.x", "debian-openssl-3.0.x", "rhel-openssl-3.0.x"]
  previewFeatures = ["fullTextSearch", "fullTextIndex", "driverAdapters"]
  output          = "../generated/prisma-tenant/client"
}
```

### Environment & setup

- OS - Latest macOS
- Database - Planetscale
- Node.js - 18

### Prisma Version

```
    "prisma": "^5.4.2",
    "@prisma/client": "^5.4.2",
```

**Comment by janpio** [comment]
Yes, I would say it's a reasonable assumption that this is a bug in our preview feature.

I would assume that somehow our raw queries do not execute on the correct database. Can you maybe try to qualify the table name `foo` with the database name `bar.foo` in the query? That might be a temporary workaround while we figure out what is going on here.

**Comment by mattrobenolt** [comment]
To add, PlanetScale typically in most cases has a correct default database to connect to.

If your string is something like `mysql://user:pass@hostname/dbname` we technically ignore the dbname since it's usually redundant.

Some drivers translate the `dbname` to a ```USE `dbname`;``` query as the first thing on the connection.

For our database-js driver, we do no implicit behavior here, and would require an explicit ```USE `mysql`;``` to get switched into that database.

I hope that helps narrow it down, but I suspect Prisma is making an assumption here and trying to explicitly ```USE `mysql`;```.

**Comment by janpio** [comment]
> I'm not actually running a raw query where it fails.

Sorry, I misread part of our stacktrace. You are correct. Removed that label.

**Comment by jkomyno** [comment]
Hi @Janushan, I'm trying to reproduce and investigate this issue. Could you please show us a couple of your (censored) database strings that you're using in this project?

E.g., do they look like `mysql://user1:password1@aws.connect.psdb.cloud/my-database-name?sslaccept=strict`?

I'm running the following
- `.env`
  ```env
  DATABASE_URL="mysql://user1:password1@aws.connect.psdb.cloud/my-database-name?sslaccept=strict"
  ```
- `schema.prisma`:
  ```prisma
  generator client {
    provider        = "prisma-client-js"
    previewFeatures = ["driverAdapters"]
  }
  
  datasource db {
    provider = "mysql"
    url      = env("DATABASE_URL")
  }
  ```
- `index.ts`:
  ```typescript
  import { connect } from '@planetscale/database'
  import { PrismaPlanetScale } from '@prisma/adapter-planetscale'
  import { PrismaClient } from '@prisma/client'

  const connectionString = `${process.env.DATABASE_URL as string}`
  
  const connection = connect({ url: connectionString })
  const adapter = new PrismaPlanetScale(connection)
  const prisma = new PrismaClient({
    adapter,
    log: [
      { level: 'query', emit: 'event' },
    ],
  })

  prisma.$on('query', (query) => {
    console.log('query', query)
  })
  
  await prisma.author.count()
  ```

and the Prisma queries (`count()` in this case) work as expected.

If I replace the `.env` file with the following - dropping the database name -, I get an error, but that's different than yours:
- `.env`
  ```env
  DATABASE_URL="mysql://user1:password1@aws.connect.psdb.cloud/?sslaccept=strict"
  ```

```
DatabaseError: syntax error at position 31
    at Connection.execute (.../prisma/prisma/reproductions/node_modules/.pnpm/@planetscale+database@1.11.0/node_modules/@planetscale/database/dist/cjs/index.js:99:19)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at PrismaPlanetScale.performIO (/Users/jkomyno/work/prisma/prisma/packages/adapter-planetscale/dist/index.js:148:22)
    at PrismaPlanetScale.queryRaw (/Users/jkomyno/work/prisma/prisma/packages/adapter-planetscale/dist/index.js:119:22)
    at <anonymous> (/Users/jkomyno/work/prisma/prisma/reproductions/driver-adapters/node_modules/.prisma/client/runtime/library.js:26:4962) {
  status: 400,
  body: { message: 'syntax error at position 31', code: 'UNKNOWN' },
  clientVersion: '0.0.0'
}
```

The error is due to Prisma generating the following query, with an empty (``) database name:
```sql
SELECT COUNT(*) FROM (SELECT ``.`authors`.`id` FROM ``.`authors` WHERE 1=1) AS `sub`
```

My hypothesis is that your setup (which manually overrides the datasource URL defined in your `schema.prisma`)

```typescript
const prisma = new PrismaClient({
    datasources: {
      db: {
        url: databaseUrl,
      },
    },
    // ...
})
```

makes Prisma connect to a URL that Prisma DID NOT use for [`prisma db push`](https://www.prisma.io/docs/concepts/components/prisma-migrate/db-push) or similar Prisma setup commands.
Can you please double-check this and get back at us with your results and a minimally reproducible project? Thanks.

---

Also, is `TenantPrismaClient` (mentioned in [your snippet above](https://github.com/prisma/prisma/issues/21511#issue-1943908799)) any different than `PrismaClient` (used in the same snippet)?

**Comment by jkomyno** [comment]
Hi @Janushan, do you have any news about [this](https://github.com/prisma/prisma/issues/21511#issuecomment-1785001796)? Thanks.

**Comment by janpio** [comment]
@Janushan Concretely, can you optimally provide a small reproduction for us of the situation you are running into?
You provided some snippets, but as you can see we could not fully figure out how you are using connection strings and then managing the driver, adapter and Prisma Client instances. A small repository that extracts this part of your project, would be optimal. Thank you!

**Comment by Janushan** [comment]
Hi @janpio and @jkomyno 
Sorry about my delay.
No problem. I'll share a repo to reproduce towards the end of this week or next.
Thanks for looking into this!

**Comment by jkomyno** [comment]
Hi @Janushan, great, looking forward to see that repro :)

**Comment by jkomyno** [comment]
Hey @Janushan, it's been a while. Can you please post the reproduction for this issue? Otherwise, we may close this issue until we have more information available. Thanks!

**Comment by jkomyno** [comment]
Closing the issue for inactivity.

**Comment by SevInf** [comment]
We haven't got a reproduction but I can make a good guess on what's happened.
In 5.7.0, connection URL is taken from the adapter, but DB name to use within generated queries still came from the schema file, even if adapter is used. If no DB specified in that URL, `mysql` prefix is used (how sensible of a default  that is, is a topic for another day).

That was the most likely cause of a problem described here - schema file url has a default connection string vwithout DB name, URL in the adapter did not affect generated queries.

We've fixed that problem in prisma/prisma-engines#4548. This fix is scheduled to go out in 5.8.0 - after that, only the URL specified within the adapter will matter and DB name will be picked up from there as well.

Workaround meanwhile: continue using `db.url` override alongside `adapter`. Note, that we are very likely to disallow this combination of options in 5.8.0 too, the same version that will contain the fix for the problem.
